src/ssl/support.cc

```diff
@@ -1140,38 +1208,42 @@ completeIssuers(X509_STORE_CTX *ctx, STACK_OF(X509) *untrustedCerts)
         debugs(83, 2,  "exceeded the maximum certificate chain length: " << depth);
 }
 
-/// OpenSSL certificate validation callback.
 static int
-untrustedToStoreCtx_cb(X509_STORE_CTX *ctx,void *data)
+squidX509VerifyCert(X509_STORE_CTX *ctx, STACK_OF(X509) *extraCerts)
 {
-    debugs(83, 4,  "Try to use pre-downloaded intermediate certificates");
-
-    SSL *ssl = static_cast<SSL *>(X509_STORE_CTX_get_ex_data(ctx, SSL_get_ex_data_X509_STORE_CTX_idx()));
-    STACK_OF(X509) *sslUntrustedStack = static_cast <STACK_OF(X509) *>(SSL_get_ex_data(ssl, ssl_ex_index_ssl_untrusted_chain));
-
     // OpenSSL already maintains ctx->untrusted but we cannot modify
     // internal OpenSSL list directly. We have to give OpenSSL our own
     // list, but it must include certificates on the OpenSSL ctx->untrusted
     STACK_OF(X509) *oldUntrusted = X509_STORE_CTX_get0_untrusted(ctx);
-    STACK_OF(X509) *sk = sk_X509_dup(oldUntrusted); // oldUntrusted is always not NULL
+    Ssl::X509_STACK_Pointer untrustedCerts(X509_chain_up_ref(oldUntrusted)); // oldUntrusted is always not nil
 
-    for (int i = 0; i < sk_X509_num(sslUntrustedStack); ++i) {
-        X509 *cert = sk_X509_value(sslUntrustedStack, i);
-        sk_X509_push(sk, cert);
+    if (extraCerts) {
+        for (int i = 0; i < sk_X509_num(extraCerts); ++i) {
+            X509 *cert = sk_X509_value(extraCerts, i);
+            X509_up_ref(cert);
+            sk_X509_push(untrustedCerts.get(), cert);
+        }
     }
 
     // If the local untrusted certificates internal database is used
     // run completeIssuers to add missing certificates if possible.
     if (SquidUntrustedCerts.size() > 0)
-        completeIssuers(ctx, sk);
+        completeIssuers(ctx, untrustedCerts.get());
 
-    X509_STORE_CTX_set0_untrusted(ctx, sk); // No locking/unlocking, just sets ctx->untrusted
+    X509_STORE_CTX_set0_untrusted(ctx, untrustedCerts.get()); // No locking/unlocking, just sets ctx->untrusted
     int ret = X509_verify_cert(ctx);
     X509_STORE_CTX_set0_untrusted(ctx, oldUntrusted); // Set back the old untrusted list
-    sk_X509_free(sk); // Release sk list
     return ret;
 }
 
+/// OpenSSL certificate validation callback.
```

Comment:
It is NOT a X509_verify_cert callback.
It is the X509_verify_cert replacement callback.  It is the function which actually verify the certificate.

The openSSL manual uses the term "verification callback function"

Reply:
> It is NOT a X509_verify_cert callback.
> It is the X509_verify_cert replacement callback.

It looks like our definitions of "callback" differ. I will try to clarify. 

Since Squid code does not call untrustedToStoreCtx_cb() directly, we should document which OpenSSL function (called by Squid) calls (directly or indirectly) untrustedToStoreCtx_cb() -- that information is not in Squid source code. OpenSSL SSL_CTX_set_cert_verify_callback() documentation does not answer this question either AFAICT. The reader has to guess the actual context/caller of this callback, which is especially tricky when we customize certificate validation. I assume you know the answer -- what it that OpenSSL function?

Reply:
We are using it with SSL_CTX_set_cert_verify_callback().
The SSL_CTX_set_cert_verify_callback actually sets the verification function, a function which return accept/not-accept for a certificate.

Reply:
You are telling me how we inform OpenSSL about untrustedToStoreCtx_cb() existence. I know that. That knowledge does not allow me to answer my untrustedToStoreCtx_cb() _usage_ or _caller_ question. I want to know which OpenSSL function (called by Squid) calls (directly or indirectly) untrustedToStoreCtx_cb().

In other words, if I assert or break inside untrustedToStoreCtx_cb(), what will be the first (i.e. smallest frame number) OpenSSL function that I will see in the backtrace? What will be the last (i.e. highest frame number) OpenSSL function that I will see in that same backtrace?

If several OpenSSL functions fit that profile, depending on Squid call context or other factors, then please provide all of them.

Reply:
Sorry I did not understand.
This callback is called while we are negotiating with the server, while the  SSL_connect is called.

> In other words, if I assert or break inside untrustedToStoreCtx_cb(), what will be the first (i.e. smallest frame number) OpenSSL function that I will see in the backtrace

The backtrace I am seeing is the following:
```
#0  squidX509VerifyCert (ctx=0x5555563b3670, extraCerts=0x0) at support.cc:1220
#1  0x00005555559c97b9 in untrustedToStoreCtx_cb (ctx=0x5555563b3670)
    at support.cc:1247
#2  0x00007ffff712f284 in ssl_verify_cert_chain (s=s@entry=0x55555663bf50, 
    sk=sk@entry=0x55555664e170) at ../ssl/ssl_cert.c:428
#3  0x00007ffff715026b in tls_process_server_certificate (s=0x55555663bf50, 
    pkt=0x7fffffffd360) at ../ssl/statem/statem_clnt.c:1906
#4  0x00007ffff7152875 in ossl_statem_client_process_message (
    s=0x55555663bf50, pkt=<optimized out>) at ../ssl/statem/statem_clnt.c:1045
#5  0x00007ffff714c39c in read_state_machine (s=0x55555663bf50)
    at ../ssl/statem/statem.c:633
#6  state_machine (s=0x55555663bf50, server=0) at ../ssl/statem/statem.c:431
#7  0x00007ffff71384c4 in SSL_do_handshake (s=0x55555663bf50)
    at ../ssl/ssl_lib.c:3570
#8  0x00005555559a07a3 in Security::PeerConnector::negotiate (
    this=0x55555663b438) at PeerConnector.cc:183
```
In the above stacktrace the SSL_connect is not shown however the SSL_connect calls the SSL_do_handshake.
Probably the backtrace  is depending on the used TLS version, maybe is a little different in other cases.

Reply:
Thank you. These details answer my question. I am leaving this change request open to remind me about polishing the untrustedToStoreCtx_cb() description.